File: src/time_to_sleep/history.py
import sqlite3
from collections.abc import Sequence
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any

from pydantic import BaseModel

from time_to_sleep.domain import UsageSnapshot


class HistoryPoint(BaseModel):
    account_id: str
    provider: str
    window_id: str
    used_percent: float
    observed_at: datetime


class HourlyUsageDistribution(BaseModel):
    hour: int
    average_percent: float
    samples_count: int


class HistoryStore:
    def __init__(self, db_path: str | Path | None = None) -> None:
        if db_path is None or db_path == ":memory:":
            self.db_path = ":memory:"
        else:
            self.db_path = str(Path(db_path).expanduser())
            Path(self.db_path).parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(self.db_path, check_same_thread=False)
        self._conn.row_factory = sqlite3.Row
        self._last_records: dict[tuple[str, str], tuple[float, datetime]] = {}
        self._last_prune: datetime = datetime.now(UTC)
        self._init_db()

    def _get_connection(self) -> sqlite3.Connection:
        return self._conn

    def _init_db(self) -> None:
        with self._get_connection() as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS usage_history (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    account_id TEXT NOT NULL,
                    provider TEXT NOT NULL,
                    window_id TEXT NOT NULL,
                    used_percent REAL NOT NULL,
                    observed_at TEXT NOT NULL
                )
                """
            )
            conn.execute(
                """
                CREATE INDEX IF NOT EXISTS idx_history_account_time 
                ON usage_history(account_id, observed_at)
                """
            )
            conn.commit()
        self.prune(max_days=30)

    def record_snapshots(self, snapshots: Sequence[UsageSnapshot]) -> None:
        now = datetime.now(UTC)
        if (now - self._last_prune).total_seconds() > 86400:
            self.prune(max_days=30)
            self._last_prune = now

        records: list[tuple[str, str, str, float, str]] = []

        for s in snapshots:
            ts_dt = s.retrieved_at or s.observed_at or now
            ts = ts_dt.isoformat()
            for w in s.windows:
                key = (s.account_id, w.id)
                if key in self._last_records:
                    last_pct, last_time = self._last_records[key]
                    if (
                        abs(w.used_percent - last_pct) < 0.001
                        and (ts_dt - last_time).total_seconds() < 300
                    ):
                        continue
                self._last_records[key] = (w.used_percent, ts_dt)
                records.append((s.account_id, s.provider, w.id, w.used_percent, ts))

        if not records:
            return

        with self._get_connection() as conn:
            conn.executemany(
                """
                INSERT INTO usage_history (
                    account_id, provider, window_id, used_percent, observed_at
                )
                VALUES (?, ?, ?, ?, ?)
                """,
                records,
            )
            conn.commit()

    def get_history(
        self,
        account_id: str | None = None,
        hours: int = 24,
    ) -> list[HistoryPoint]:
        cutoff = (datetime.now(UTC) - timedelta(hours=hours)).isoformat()
        query = """
            SELECT account_id, provider, window_id, used_percent, observed_at
            FROM usage_history
            WHERE observed_at >= ?
        """
        params: list[Any] = [cutoff]

        if account_id:
            query += " AND account_id = ?"
            params.append(account_id)

        query += " ORDER BY observed_at ASC"

        with self._get_connection() as conn:
            cursor = conn.execute(query, params)
            rows = cursor.fetchall()

        points: list[HistoryPoint] = []
        for r in rows:
            points.append(
                HistoryPoint(
                    account_id=r["account_id"],
                    provider=r["provider"],
                    window_id=r["window_id"],
                    used_percent=float(r["used_percent"]),
                    observed_at=datetime.fromisoformat(r["observed_at"]),
                )
            )
        return points

    def get_hourly_heatmap(
        self, account_id: str | None = None, days: int = 7
    ) -> list[HourlyUsageDistribution]:
        cutoff = (datetime.now(UTC) - timedelta(days=days)).isoformat()
        query = """
            SELECT CAST(strftime('%H', observed_at) AS INTEGER) AS hr,
                   AVG(used_percent) AS avg_pct,
                   COUNT(*) AS sample_cnt
            FROM usage_history
            WHERE observed_at >= ?
        """
        params: list[Any] = [cutoff]
        if account_id:
            query += " AND account_id = ?"
            params.append(account_id)

        query += " GROUP BY hr"

        with self._get_connection() as conn:
            cursor = conn.execute(query, params)
            rows = {row["hr"]: (row["avg_pct"], row["sample_cnt"]) for row in cursor.fetchall()}

        result: list[HourlyUsageDistribution] = []
        for h in range(24):
            if h in rows:
                avg, count = rows[h]
                result.append(
                    HourlyUsageDistribution(
                        hour=h,
                        average_percent=round(float(avg), 1),
                        samples_count=int(count),
                    )
                )
            else:
                result.append(
                    HourlyUsageDistribution(
                        hour=h,
                        average_percent=0.0,
                        samples_count=0,
                    )
                )
        return result

    def prune(self, max_days: int = 30) -> int:
        cutoff = (datetime.now(UTC) - timedelta(days=max_days)).isoformat()
        with self._get_connection() as conn:
            cursor = conn.execute("DELETE FROM usage_history WHERE observed_at < ?", (cutoff,))
            deleted = cursor.rowcount
            conn.commit()
        return deleted


_default_history_path = Path("~/.config/time-to-sleep/history.db").expanduser()
default_history_store = HistoryStore(_default_history_path)

File: src/history.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

use anyhow::{Context, Result};
use chrono::{DateTime, Duration, Utc};
use rusqlite::{params, Connection};

use crate::domain::{HistoryPoint, HourlyUsageDistribution, UsageSnapshot};

type SnapshotCache = HashMap<(String, String), (f64, DateTime<Utc>)>;

pub struct HistoryStore {
    conn: Mutex<Connection>,
    last_records: Mutex<SnapshotCache>,
    last_prune: Mutex<DateTime<Utc>>,
}

impl HistoryStore {
    pub fn new(db_path: Option<&Path>) -> Result<Self> {
        let conn = match db_path {
            None => Connection::open_in_memory()?,
            Some(p) if p.to_str() == Some(":memory:") => Connection::open_in_memory()?,
            Some(p) => {
                if let Some(parent) = p.parent() {
                    std::fs::create_dir_all(parent)?;
                }
                Connection::open(p)?
            }
        };

        // SQLite speed pragmas
        let _ = conn.execute_batch("PRAGMA journal_mode = WAL; PRAGMA synchronous = NORMAL;");

        let store = Self {
            conn: Mutex::new(conn),
            last_records: Mutex::new(HashMap::new()),
            last_prune: Mutex::new(Utc::now()),
        };

        store.init_db()?;
        let _ = store.prune(30);

        Ok(store)
    }

    pub fn default_path() -> PathBuf {
        if let Some(home) = dirs::home_dir() {
            home.join(".config/time-to-sleep/history.db")
        } else {
            PathBuf::from("history.db")
        }
    }

    fn init_db(&self) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute_batch(
            "
            CREATE TABLE IF NOT EXISTS usage_history (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                account_id TEXT NOT NULL,
                provider TEXT NOT NULL,
                window_id TEXT NOT NULL,
                used_percent REAL NOT NULL,
                observed_at TEXT NOT NULL
            );
            CREATE INDEX IF NOT EXISTS idx_history_account_time
            ON usage_history(account_id, observed_at);
            CREATE INDEX IF NOT EXISTS idx_history_account_window_time
            ON usage_history(account_id, window_id, observed_at);
            ",
        )?;
        Ok(())
    }

    pub fn record_snapshots(&self, snapshots: &[UsageSnapshot]) -> Result<()> {
        let now = Utc::now();
        {
            let mut last_prune = self.last_prune.lock().unwrap();
            if (now - *last_prune).num_seconds() > 86400 {
                let _ = self.prune(30);
                *last_prune = now;
            }
        }

        let mut last_records = self.last_records.lock().unwrap();
        let mut pending_cache_updates: SnapshotCache = HashMap::new();
        let mut to_insert = Vec::new();

        for snapshot in snapshots {
            // Persist the time at which provider data was observed. Cached retrieval time is
            // only a fallback when the provider did not supply an observation timestamp.
            let ts = snapshot
                .observed_at
                .or(snapshot.retrieved_at)
                .unwrap_or(now);
            let ts_str = ts.to_rfc3339();

            for window in &snapshot.windows {
                let key = (snapshot.account_id.clone(), window.id.clone());
                let previous = pending_cache_updates
                    .get(&key)
                    .or_else(|| last_records.get(&key));

                if let Some((prev_pct, prev_time)) = previous {
                    let elapsed = (ts - *prev_time).num_seconds();
                    if elapsed < 0 {
                        // Do not let an out-of-order cached observation move history backward.
                        continue;
                    }
                    if (window.used_percent - *prev_pct).abs() < 0.001 && elapsed < 300 {
                        continue;
                    }
                }

                pending_cache_updates.insert(key, (window.used_percent, ts));
                to_insert.push((
                    snapshot.account_id.clone(),
                    snapshot.provider.as_str().to_string(),
                    window.id.clone(),
                    window.used_percent,
                    ts_str.clone(),
                ));
            }
        }

        if to_insert.is_empty() {
            return Ok(());
        }

        let mut conn = self.conn.lock().unwrap();
        let tx = conn.transaction()?;
        {
            let mut stmt = tx.prepare_cached(
                "INSERT INTO usage_history (account_id, provider, window_id, used_percent, observed_at) VALUES (?, ?, ?, ?, ?)"
            )?;
            for (account_id, provider, window_id, used_percent, observed_at) in &to_insert {
                stmt.execute(params![
                    account_id,
                    provider,
                    window_id,
                    used_percent,
                    observed_at
                ])?;
            }
        }
        tx.commit()?;

        // The dedup cache represents successfully persisted state. Updating it only after the
        // transaction commits ensures a failed write can be retried immediately.
        last_records.extend(pending_cache_updates);
        Ok(())
    }

    fn history_point_from_row(row: &rusqlite::Row<'_>) -> Result<HistoryPoint> {
        let ts_str: String = row.get(4)?;
        let observed_at = DateTime::parse_from_rfc3339(&ts_str)
            .with_context(|| format!("invalid observed_at value in usage_history: {ts_str}"))?
            .with_timezone(&Utc);

        Ok(HistoryPoint {
            account_id: row.get(0)?,
            provider: row.get(1)?,
            window_id: row.get(2)?,
            used_percent: row.get(3)?,
            observed_at,
        })
    }

    pub fn get_history(&self, account_id: Option<&str>, hours: i64) -> Result<Vec<HistoryPoint>> {
        let cutoff = (Utc::now() - Duration::hours(hours)).to_rfc3339();
        let conn = self.conn.lock().unwrap();

        let mut query = "SELECT account_id, provider, window_id, used_percent, observed_at FROM usage_history WHERE observed_at >= ?".to_string();
        if account_id.is_some() {
            query.push_str(" AND account_id = ?");
        }
        query.push_str(" ORDER BY observed_at ASC");

        let mut stmt = conn.prepare(&query)?;
        let mut points = Vec::new();

        if let Some(acc_id) = account_id {
            let mut rows = stmt.query(params![cutoff, acc_id])?;
            while let Some(row) = rows.next()? {
                points.push(Self::history_point_from_row(row)?);
            }
        } else {
            let mut rows = stmt.query(params![cutoff])?;
            while let Some(row) = rows.next()? {
                points.push(Self::history_point_from_row(row)?);
            }
        }

        Ok(points)
    }

    pub fn get_hourly_heatmap(
        &self,
        account_id: Option<&str>,
        days: i64,
    ) -> Result<Vec<HourlyUsageDistribution>> {
        let cutoff = (Utc::now() - Duration::days(days)).to_rfc3339();
        let conn = self.conn.lock().unwrap();

        let mut query = "SELECT CAST(strftime('%H', observed_at) AS INTEGER) AS hr, AVG(used_percent) AS avg_pct, COUNT(*) AS sample_cnt FROM usage_history WHERE observed_at >= ?".to_string();
        if account_id.is_some() {
            query.push_str(" AND account_id = ?");
        }
        query.push_str(" GROUP BY hr");

        let mut stmt = conn.prepare(&query)?;
        let mut map: HashMap<i32, (f64, i64)> = HashMap::new();

        if let Some(acc_id) = account_id {
            let mut rows = stmt.query(params![cutoff, acc_id])?;
            while let Some(row) = rows.next()? {
                let hr: i32 = row.get(0)?;
                let avg: f64 = row.get(1)?;
                let count: i64 = row.get(2)?;
                map.insert(hr, (avg, count));
            }
        } else {
            let mut rows = stmt.query(params![cutoff])?;
            while let Some(row) = rows.next()? {
                let hr: i32 = row.get(0)?;
                let avg: f64 = row.get(1)?;
                let count: i64 = row.get(2)?;
                map.insert(hr, (avg, count));
            }
        };

        let mut result = Vec::with_capacity(24);
        for h in 0..24 {
            if let Some(&(avg, count)) = map.get(&h) {
                result.push(HourlyUsageDistribution {
                    hour: h,
                    average_percent: (avg * 10.0).round() / 10.0,
                    samples_count: count,
                });
            } else {
                result.push(HourlyUsageDistribution {
                    hour: h,
                    average_percent: 0.0,
                    samples_count: 0,
                });
            }
        }

        Ok(result)
    }

    pub fn prune(&self, max_days: i64) -> Result<usize> {
        let cutoff = (Utc::now() - Duration::days(max_days)).to_rfc3339();
        let conn = self.conn.lock().unwrap();
        let deleted = conn
            .execute(
                "DELETE FROM usage_history WHERE observed_at < ?",
                params![cutoff],
            )
            .context("Failed to prune usage history")?;
        Ok(deleted)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::{AccountStatus, ErrorCode, ProviderName, UsageWindow};

    #[test]
    fn test_history_dedup_and_heatmap() {
        let store = HistoryStore::new(None).unwrap();
        let now = Utc::now();

        let snap = UsageSnapshot {
            account_id: "codex-primary".to_string(),
            provider: ProviderName::Codex,
            configured_email: "test@example.com".to_string(),
            observed_email: None,
            status: AccountStatus::Live,
            error_code: ErrorCode::None,
            message: None,
            source: "test".to_string(),
            plan_type: None,
            observed_at: Some(now),
            retrieved_at: Some(now),
            windows: vec![UsageWindow {
                id: "primary".to_string(),
                used_percent: 50.0,
                window_minutes: Some(300),
                resets_at: None,
                raw_limits: None,
            }],
        };

        // First insert
        store.record_snapshots(&[snap.clone()]).unwrap();
        // Duplicate should be skipped
        store.record_snapshots(&[snap.clone()]).unwrap();

        let history = store.get_history(Some("codex-primary"), 24).unwrap();
        assert_eq!(history.len(), 1);

        let heatmap = store.get_hourly_heatmap(Some("codex-primary"), 7).unwrap();
        assert_eq!(heatmap.len(), 24);
        let hr = now.format("%H").to_string().parse::<i32>().unwrap();
        let entry = heatmap.iter().find(|h| h.hour == hr).unwrap();
        assert_eq!(entry.average_percent, 50.0);
        assert_eq!(entry.samples_count, 1);
    }
}
